### PR TITLE
Impl data jpa

### PR DESCRIPTION
**Pull request by ConfusedSatlan**

### CODE REVIEW

File: src/main/java/application/onlinebookstore/repository/book/spec/AuthorSpecificationProvider.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
+package application.onlinebookstore.repository.book.spec;
+
+import application.onlinebookstore.model.Book;
+import application.onlinebookstore.repository.SpecificationProvider;
+import java.util.Arrays;
+import org.springframework.data.jpa.domain.Specification;
+
+public class AuthorSpecificationProvider implements SpecificationProvider<Book> {
```
Comment (keda-vlad) [inline src/main/java/application/onlinebookstore/repository/book/spec/AuthorSpecificationProvider.java]: U need annotate this for creating bean

File: src/main/java/application/onlinebookstore/repository/book/spec/TitleSpecificationProvider.java [hunk 2]
```diff
@@ -0,0 +1,19 @@
+package application.onlinebookstore.repository.book.spec;
+
+import application.onlinebookstore.model.Book;
+import application.onlinebookstore.repository.SpecificationProvider;
+import java.util.Arrays;
+import org.springframework.data.jpa.domain.Specification;
+
+public class TitleSpecificationProvider implements SpecificationProvider<Book> {
```
Comment (keda-vlad) [inline src/main/java/application/onlinebookstore/repository/book/spec/TitleSpecificationProvider.java]: U need annotate this for creating bean

File: src/main/java/application/onlinebookstore/consthelper/ArraysConst.java [hunk 3]
```diff
@@ -0,0 +1,5 @@
+package application.onlinebookstore.consthelper;
+
+public class ArraysConst {
+    public static final int ZERO_LENGTH = 0;
+}
```
Comment (keda-vlad) [inline src/main/java/application/onlinebookstore/consthelper/ArraysConst.java]: Why do u need this?

File: src/main/java/application/onlinebookstore/controller/BookController.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.bind.annotation.RestController;
+
+@RequiredArgsConstructor
+@RestController
+@RequestMapping(value = "/api/books")
+public class BookController {
+    private final BookService bookService;
```
Comment (keda-vlad) [inline src/main/java/application/onlinebookstore/controller/BookController.java:25]: Book Endpoints:

GET: /api/books (Retrieve book catalog) (should be done in the previous PR)
GET: /api/books/{id} (Retrieve book details) (should be done in the previous PR)
POST: /api/books (Create a new book) (should be done in the previous PR)
PUT: /api/books/{id} (Update a specific book) <----------------------------------------- U miss this

File: src/main/java/application/onlinebookstore/repository/book/impl/BookSpecificationBuilder.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+public class BookSpecificationBuilder implements SpecificationBuilder<Book> {
+    private SpecificationProviderManager<Book> specificationProviderManager;
+
+    @Override
+    public Specification<Book> build(BookSearchParametersDto searchParameters) {
+        Specification<Book> spec = Specification.where(null);
+        String[] titles = searchParameters.titles();
+        if (titles != null && titles.length > ArraysConst.ZERO_LENGTH) {
+            spec = spec.and(specificationProviderManager
+                    .getSpecificationProvider("title")
```
Comment (keda-vlad) [inline src/main/java/application/onlinebookstore/repository/book/impl/BookSpecificationBuilder.java]: It`s better to make constant for this, and in all same case

File: src/main/java/application/onlinebookstore/repository/book/spec/AuthorSpecificationProvider.java [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+import org.springframework.data.jpa.domain.Specification;
+
+public class AuthorSpecificationProvider implements SpecificationProvider<Book> {
+    @Override
+    public String getKey() {
+        return "author";
+    }
+
+    public Specification<Book> getSpecification(String[] param) {
+        return (root, query, criteriaBuilder) -> root.get("author")
```
Comment (keda-vlad) [inline src/main/java/application/onlinebookstore/repository/book/spec/AuthorSpecificationProvider.java]: It`s better to make constant for this, and in all same case

**Review by keda-vlad — CHANGES_REQUESTED** [review]

**Review by keda-vlad — APPROVED** [review]

File: src/main/java/application/onlinebookstore/service/impl/BookServiceImpl.java [hunk 7]
```diff
@@ -1,28 +1,91 @@
...
+    @Override
+    public BookDto getBookById(Long id) {
+        return bookMapper.toDto(bookRepository.findById(id)
+                .orElseThrow(
+                        () -> new EntityNotFoundException("Can't find book by id: " + id)
+                ));
+    }
+
+    @Override
+    public List<BookDto> getBookByTitle(String title) {
```
Comment (vovasalyha) [inline src/main/java/application/onlinebookstore/service/impl/BookServiceImpl.java]: Method is called getBook... but you actually returning a list, seems confusing. Whether call it getBooks... or return a single book.

File: src/main/java/application/onlinebookstore/service/impl/BookServiceImpl.java [hunk 8]
```diff
@@ -1,28 +1,91 @@
...
+        List<Book> allByTitleContainsIgnoreCase = bookRepository
+                .findAllByTitleContainsIgnoreCase(title);
+        if (allByTitleContainsIgnoreCase.isEmpty()) {
```
Comment (vovasalyha) [inline src/main/java/application/onlinebookstore/service/impl/BookServiceImpl.java]: if returning a list you don't need to throw an exception if it's empty, returning an empty list is OK

File: src/main/java/application/onlinebookstore/service/impl/BookServiceImpl.java [hunk 9]
```diff
@@ -1,28 +1,91 @@
...
+            throw new EntityNotFoundException("Can't find book by title: " + title);
+        }
+        return allByTitleContainsIgnoreCase.stream()
+                .map(bookMapper::toDto)
+                .toList();
+    }
+
+    @Override
+    public List<BookDto> getBookByAuthor(String author) {
```
Comment (vovasalyha) [inline src/main/java/application/onlinebookstore/service/impl/BookServiceImpl.java]: the same as for title

**Review by vovasalyha — CHANGES_REQUESTED** [review]
Good. Please take a look on my comments.

**Review by Vasul-Pankiv — APPROVED** [review]

**Pull request merged**